cluster/cluster_test.go

```diff
@@ -556,58 +556,202 @@ var _ = Describe("cluster/cluster tests", func() {
+			expectPathToExist(path.Join(testDir, "foo"))
+			Expect(clusterOutput.NumErrors).To(Equal(0))
+			Expect(clusterOutput.FailedCommands).To(HaveLen(0))
+			Expect(clusterOutput.RetriedCommands).To(HaveLen(1))
+			Expect(clusterOutput.RetriedCommands[0].RetryError.Error()).To(Equal("attempt 1: error was exit status 1: \nattempt 2: error was exit status 1: "))
+		})
+		It("retries a command until it reaches max retries", func() {
+			testCluster := cluster.Cluster{}
+			commandList := []cluster.ShellCommand{
+				cluster.NewShellCommand(cluster.ON_SEGMENTS|cluster.INCLUDE_COORDINATOR, -1, "", []string{"touch", path.Join(testDir, "foo")}),
+				cluster.NewShellCommand(cluster.ON_SEGMENTS|cluster.INCLUDE_COORDINATOR, 0, "", []string{"some-non-existent-command"}),
 			}
+			testCluster.Executor = &cluster.GPDBExecutor{}
+			clusterOutput := testCluster.ExecuteClusterCommandWithRetries(cluster.ON_SEGMENTS|cluster.INCLUDE_COORDINATOR, commandList, 3, 5*time.Millisecond)
+			expectedErrMsg := "exec: \"some-non-existent-command\": executable file not found in $PATH"
+			expectPathToExist(path.Join(testDir, "foo"))
+			Expect(clusterOutput.NumErrors).To(Equal(1))
+			Expect(clusterOutput.FailedCommands).To(HaveLen(1))
+			Expect(clusterOutput.RetriedCommands).To(HaveLen(0))
+			Expect(clusterOutput.FailedCommands[0].Error.Error()).To(Equal(expectedErrMsg))
+			Expect(clusterOutput.FailedCommands[0].RetryError.Error()).To(Equal(fmt.Sprintf("attempt 1: error was %s: \nattempt 2: error was %s: \nattempt 3: error was %s: ", expectedErrMsg, expectedErrMsg, expectedErrMsg)))
 		})
-		DescribeTable("CheckClusterError", func(scope cluster.Scope, includeCoordinator bool, perSegment bool, remote bool) {
-			remoteOutput.Scope = scope
-			remoteOutput.Commands[0].Scope = scope
-			remoteOutput.FailedCommands[0].Scope = scope
-			errStr := "1 segment"
-			debugStr := "segment 1 on host remotehost1"
-			var generatorFunc interface{}
-			generatorFunc = func(contentID int) string { return "Error received" }
-			if !perSegment {
-				errStr = "1 host"
-				debugStr = "host remotehost1"
+	})
+	Describe("CheckClusterError", func() {
+		Context("FailedCommands", func() {
+			var (
+				remoteOutput *cluster.RemoteOutput
+				failedCmd    cluster.ShellCommand
+			)
+			BeforeEach(func() {
+				failedCmd = cluster.ShellCommand{
+					Scope:         0, // The appropriate scope will be set in each test
+					Content:       1,
+					Host:          "remotehost1",
+					Command:       nil,
+					CommandString: "this is the command",
+					Stderr:        "exit status 1",
+					Error:         fmt.Errorf("command error"),
+				}
+				remoteOutput = &cluster.RemoteOutput{
+					Scope:          0,
+					NumErrors:      1,
+					Commands:       []cluster.ShellCommand{failedCmd},
+					FailedCommands: []cluster.ShellCommand{failedCmd},
+				}
+			})
+			DescribeTable("CheckClusterError", func(scope cluster.Scope, includeCoordinator bool, perSegment bool, remote bool) {
```

Comment:
`includeCoordinator` looks unused in this test. We can remove?

Reply:
Interesting. I see we passed different values below so I think it was intended to be used. Maybe I dropped part of the test by mistake or something.

Reply:
I checked and it seems like all the information the functions needs about that is included in scope. I removed the test variable